RecoJets/JetProducers/src/PileupJetIdAlgo.cc

```diff
@@ -382,67 +387,97 @@ PileupJetIdentifier PileupJetIdAlgo::computeIdVariables(const reco::Jet * jet, f
-			try { 
-				float tkpt = icand->trackRef()->pt(); 
-				sumTkPt += tkpt;
-				// 'classic' beta definition based on track-vertex association
-				bool inVtx0 = find( vtx->tracks_begin(), vtx->tracks_end(), reco::TrackBaseRef(icand->trackRef()) ) != vtx->tracks_end();
-				bool inAnyOther = false;
-				// alternative beta definition based on track-vertex distance of closest approach
-				double dZ0 = fabs(icand->trackRef()->dz(vtx->position()));
-				double dZ = dZ0;
-				for(reco::VertexCollection::const_iterator  vi=allvtx.begin(); vi!=allvtx.end(); ++vi ) {
-					const reco::Vertex & iv = *vi;
-					if( iv.isFake() || iv.ndof() < 4 ) { continue; }
-					// the primary vertex may have been copied by the user: check identity by position
-					bool isVtx0  = (iv.position() - vtx->position()).r() < 0.02;
-					// 'classic' beta definition: check if the track is associated with any vertex other than the primary one
-					if( ! isVtx0 && ! inAnyOther ) {
-						inAnyOther = find( iv.tracks_begin(), iv.tracks_end(), reco::TrackBaseRef(icand->trackRef()) ) != iv.tracks_end();
+		// // beta and betastar		
+		if(  icand->charge() != 0 ) {
+		      if (!isPacked){
+				try { 
+	      			float tkpt = candPt; //icand->trackRef()->pt();
+					sumTkPt += tkpt;
+					// 'classic' beta definition based on track-vertex association
+					bool inVtx0 = find( vtx->tracks_begin(), vtx->tracks_end(), reco::TrackBaseRef(lPF->trackRef()) ) != vtx->tracks_end();
+					bool inAnyOther = false;
+					// alternative beta definition based on track-vertex distance of closest approach
+					double dZ0 = std::abs(lPF->trackRef()->dz(vtx->position()));
+					double dZ = dZ0;
+					for(reco::VertexCollection::const_iterator  vi=allvtx.begin(); vi!=allvtx.end(); ++vi ) {
+						const reco::Vertex & iv = *vi;
+						if( iv.isFake() || iv.ndof() < 4 ) { continue; }
+						// the primary vertex may have been copied by the user: check identity by position
+						bool isVtx0  = (iv.position() - vtx->position()).r() < 0.02;
+						// 'classic' beta definition: check if the track is associated with any vertex other than the primary one
+						if( ! isVtx0 && ! inAnyOther ) {
+							inAnyOther = find( iv.tracks_begin(), iv.tracks_end(), reco::TrackBaseRef(lPF->trackRef()) ) != iv.tracks_end();
+						}
+						// alternative beta: find closest vertex to the track
+						dZ = std::min(dZ,std::abs(lPF->trackRef()->dz(iv.position())));
 					}
-					// alternative beta: find closest vertex to the track
-					dZ = std::min(dZ,fabs(icand->trackRef()->dz(iv.position())));
+					// classic beta/betaStar
+					if( inVtx0 && ! inAnyOther ) {
+						internalId_.betaClassic_ += tkpt;
+					} else if( ! inVtx0 && inAnyOther ) {
+						internalId_.betaStarClassic_ += tkpt;
+					}
+					// alternative beta/betaStar
+					if( dZ0 < 0.2 ) {
+						internalId_.beta_ += tkpt;
+					} else if( dZ < 0.2 ) {
+						internalId_.betaStar_ += tkpt;
+					}
+				} catch (cms::Exception &e) {
```

Comment:
Sometimes exceptions due to invalid references on tracks linked to the candidates are raised by the try / catch statement. As it is not a showstopper for the algorithm (if a jet contains several tracks), it was displayed as a warning rather than failing.
Keeping a warning would be the best, otr at least a summary of them at the end of the job, rather than removing the try/catch statement.
@nhanvtran , what do you prefer?